dawn/src/dawn/Support/IteratorAdapters.h

```diff
@@ -0,0 +1,95 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_SUPPORT_ITERATORADAPTERS_H
+#define DAWN_SUPPORT_ITERATORADAPTERS_H
+
+#include <iterator>
+#include <numeric>
+#include <tuple>
+
+namespace dawn {
+
+// enumerate on containers
+template <typename T, typename TIter = decltype(std::begin(std::declval<T>())),
+          typename = decltype(std::end(std::declval<T>()))>
+constexpr auto enumerate(T&& iterable) {
+  struct iterator {
+    size_t i;
+    TIter iter;
+    bool operator!=(const iterator& other) const { return iter != other.iter; }
+    void operator++() {
+      ++i;
+      ++iter;
+    }
+    auto operator*() const { return std::tie(i, *iter); }
+  };
+  struct iterable_wrapper {
+    T iterable;
+    auto begin() { return iterator{0, std::begin(iterable)}; }
+    auto end() { return iterator{0, std::end(iterable)}; }
+  };
+  return iterable_wrapper{std::forward<T>(iterable)};
+}
+
+// reverse
+template <typename T>
+struct reversion_wrapper {
+  T& iterable;
+};
+
+template <typename T>
+auto begin(reversion_wrapper<T> w) {
```

Comment:
I would rather define `begin(reversion_wrapper<T>&)` and `begin(reversion_wrapper<T> const &)`.